src/io-msbc.c

```diff
@@ -0,0 +1,253 @@
+{
+	size_t written = 0;
+
+	if (sbc->enc_buffer_cnt < length) {
+		warn("Encoded data underflow");
+		return -1;
+	}
+
+	if ((written = write(bt_fd, sbc->enc_buffer, length)) == -1) {
+		if (errno != EWOULDBLOCK && errno != EAGAIN)
+			warn("Could not write to mSBC socket: %s", strerror(errno));
+		return -1;
+	}
+
+	memmove(sbc->enc_buffer,
+		sbc->enc_buffer + written,
+		sbc->enc_buffer_cnt - written);
+
+	sbc->enc_buffer_cnt -= written;
+
+	return 0;
+}
+
+static void iothread_encode_msbc_frames(struct sbc_state *sbc)
+{
+	static const uint8_t h2_header_frame_number[] = {
+		0x08, 0x38, 0xc8, 0xf8
+	};
+
+	size_t written, pcm_consumed = 0;
+	ssize_t len;
+
+	/* Encode all we can */
+	while ((sbc->enc_pcm_buffer_cnt - pcm_consumed) >= sbc->enc_pcm_size &&
+	       (sbc->enc_buffer_size - sbc->enc_buffer_cnt) >= SCO_H2_FRAME_LEN) {
+
+		struct msbc_frame *frame = (struct msbc_frame*)(sbc->enc_buffer
+								+ sbc->enc_buffer_cnt);
+
+		if ((len = sbc_encode(&sbc->enc,
+				      sbc->enc_pcm_buffer + pcm_consumed,
+				      sbc->enc_pcm_buffer_cnt - pcm_consumed,
+				      frame->payload,
+				      sizeof(frame->payload),
+				      &written)) < 0) {
+			error("Unable to encode mSBC: %s", strerror(-len));
+			return;
+		};
+
+		pcm_consumed += len;
+
+		frame->h2_header[0] = SCO_H2_HDR_0;
+		frame->h2_header[1] = h2_header_frame_number[sbc->enc_frame_number];
+		sbc->enc_frame_number = ((sbc->enc_frame_number) + 1) % 4;
+		sbc->enc_buffer_cnt += sizeof(*frame);
+
+#ifdef SILENCE
+		memcpy(frame->payload, msbc_zero, sizeof(frame->payload));
+#endif
+	}
+	/* Reshuffle remaining PCM samples to the beginning of the buffer */
+	memmove(sbc->enc_pcm_buffer,
+		sbc->enc_pcm_buffer + pcm_consumed,
+		sbc->enc_pcm_buffer_cnt - pcm_consumed);
+
+	/* And deduct consumed data */
+	sbc->enc_pcm_buffer_cnt -= pcm_consumed;
+}
+
+static void iothread_find_and_decode_msbc(int pcm_fd, struct sbc_state *sbc)
+{
+	ssize_t len;
+	size_t bytes_left = sbc->dec_buffer_cnt;
+	uint8_t *p = (uint8_t*) sbc->dec_buffer;
+
+	/* Find frame start */
+	while (bytes_left >= (SCO_H2_HDR_LEN + sbc->sbc_frame_len)) {
+		if (p[0] == SCO_H2_HDR_0 && p[2] == MSBC_SYNC) {
+			/* Found frame.  TODO: Check SEQ, implement PLC */
+			size_t decoded = 0;
+			if ((len = sbc_decode(&sbc->dec,
+					      p + 2,
+					      sbc->sbc_frame_len,
+					      sbc->dec_pcm_buffer,
+					      sizeof(sbc->dec_pcm_buffer),
+					      &decoded)) < 0) {
+				error("mSBC decoding error: %s\n", strerror(-len));
+				sbc->dec_buffer_cnt = 0;
+				return;
+			}
+			bytes_left -= len + SCO_H2_HDR_LEN;
+			p += len + SCO_H2_HDR_LEN;
+			if (write(pcm_fd, sbc->dec_pcm_buffer, decoded) < 0)
+				warn("Could not write PCM data: %s", strerror(errno));
+		}
+		else {
+			bytes_left--;
+			p++;
+		}
+	}
+	memmove(sbc->dec_buffer, p, bytes_left);
+	sbc->dec_buffer_cnt = bytes_left;
+}
+
+struct sbc_state* iothread_initialize_msbc(struct sbc_state *sbc)
+{
+	if (!sbc) {
+		sbc = malloc(sizeof(*sbc));
```

Comment:
Why not using memory allocated by the caller? It would be more consistent with the sbc library. And the size of the sbc_state structure is constant, so one might allocate it on the stack.